### PR TITLE
-O0 Hello world working with MINIMAL_RUNTIME.

### PR DESCRIPTION
**Pull request by juj**
This makes a first step towards getting MINIMAL_RUNTIME working with wasm backend.

### CODE REVIEW

File: tests/test_core.py [hunk 1]
```diff
@@ -8142,12 +8142,19 @@ def test_minimal_runtime_no_declare_asm_module_exports(self):
 
   # Tests that -s MINIMAL_RUNTIME=1 works well
   @no_emterpreter
-  @no_wasm_backend('MINIMAL_RUNTIME not yet available in Wasm backend')
+  @no_asan('TODO: ASan with MINIMAL_RUNTIME')
   def test_minimal_runtime_hello_world(self):
+    if self.get_setting('WASM_BACKEND') and not self.get_setting('WASM'):
```
Comment (kripken) [inline tests/test_core.py]: There is a `@no_wasm2js` decorator.

File: tests/test_core.py [hunk 2]
```diff
@@ -8142,12 +8142,19 @@ def test_minimal_runtime_no_declare_asm_module_exports(self):
...
+      return self.skipTest('TODO: wasm2js support not yet available with MINIMAL_RUNTIME')
+    if '-O2' in self.emcc_args or '-O3' in self.emcc_args or '-Os' in self.emcc_args or '-Oz' in self.emcc_args or '-fsanitize=leak' in self.emcc_args:
+      return self.skipTest('TODO: -O2 and higher with wasm backend')
+    if '-fsanitize=leak' in self.emcc_args:
```
Comment (kripken) [inline tests/test_core.py]: maybe let's add a `@no_lsan`? or `@no_san` to disable all sanitizers?

File: src/postamble_minimal.js [hunk 3]
```diff
@@ -77,10 +77,29 @@ WebAssembly.instantiate(Module['wasm'], imports).then(function(output) {
   var asm = output.instance.exports;
 #if DECLARE_ASM_MODULE_EXPORTS == 0
 
+#if WASM_BACKEND
+  // XXX Hack: some function names need to be mangled when exporting them from wasm module, others do not. 
+  // https://github.com/emscripten-core/emscripten/issues/10054
+  // Keep in sync with emscripten.py function treat_as_user_function(name).
```
Comment (kripken) [inline src/postamble_minimal.js]: I think we can add the array of names here to `settings_internal.js`, which is available both in `emscripten.py` and here.

File: emscripten.py [hunk 4]
```diff
@@ -2623,8 +2644,36 @@ def create_receiving_wasm(exports):
...
+      if shared.Settings.WASM and shared.Settings.MINIMAL_RUNTIME:
+        # In Wasm exports are assigned inside a function to variables existing in top level JS scope, i.e.
+        # var _main;
+        # WebAssembly.instantiate(Module["wasm"], imports).then((function(output) {
+        # var asm = output.instance.exports;
+        # _main = asm["_main"];
+        receiving += [asmjs_mangle(s) + ' = asm["' + s + '"];' for s in exports]
+      else:
+        if shared.Settings.MINIMAL_RUNTIME:
+          # In asm.js exports can be directly processed at top level, i.e.
```
Comment (kripken) [inline emscripten.py]: this is in wasm backend code, so asm.js should be wasm2js I think?

File: emscripten.py [hunk 5]
```diff
@@ -2623,8 +2644,36 @@ def create_receiving_wasm(exports):
 ''' % {'mangled': asmjs_mangle(e), 'e': e, 'assertions': runtime_assertions})
 
   if not shared.Settings.SWAPPABLE_ASM_MODULE:
-    for e in exports:
-      receiving.append('var %(mangled)s = Module["%(mangled)s"] = asm["%(e)s"];' % {'mangled': asmjs_mangle(e), 'e': e})
+    if shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
+      if shared.Settings.WASM and shared.Settings.MINIMAL_RUNTIME:
```
Comment (kripken) [inline emscripten.py]: why does wasm2js output differ from wasm output here? in general in the wasm backend the js/wasm paths are much more similar than in fastcomp (wasm2js literally polyfills the WebAssembly APIs, and everything should be identical except for that), so I was hoping we would not have differences here.

Comment (juj) [reply emscripten.py]: I don't know yet. I have not yet tried against wasm2js. I just put this like this to follow the fastcomp route, and did not go too far making modifications to keep the amount to review small for each PR.

**Comment by juj** [comment]
Check - updated to address review and pass tests, I'll land this once tests come through, and proceed to next set of features.

File: src/parseTools.js [hunk 6]
```diff
@@ -1631,3 +1631,12 @@ function makeRemovedFSAssert(fsName) {
   if (SYSTEM_JS_LIBRARIES.indexOf('library_' + lower + '.js') >= 0) return '';
   return "var " + fsName + " = '" + fsName + " is no longer included by default; build with -l" + lower + ".js';";
 }
+
+// Given an array of elements [elem1,elem2,elem3], returns a string "['elem1','elem2','elem3']"
+function buildStringArray(array) {
+  if (array.length > 0) {
+    return "['" + array.join("','") + "']";
+  } else {
+    return [];
```
Comment (kripken) [inline src/parseTools.js]: shouldn't this be a string too?

Comment (juj) [reply src/parseTools.js]: That would certainly help the code to work..

**Pull request merged**